File: core/ConstructBinaryTreeFromPreorderAndInorderTraversal/ConstructBinaryTreeFromPreorderAndInorderTraversal.java
package com.dbc;

public class ConstructBinaryTreeFromPreorderAndInorderTraversal {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public TreeNode dfs(int[] preorder, int[] inorder, int preLeft, int preRight, int inLeft, int inRight) {
        TreeNode root = new TreeNode(preorder[preLeft]);
        int point = inLeft;
        while (inorder[point] != preorder[preLeft]) point++;
        if (point > inLeft) {
            root.left = dfs(preorder, inorder, preLeft + 1, preLeft + point - inLeft, inLeft, point - 1);
        }
        if (point < inRight) {
            root.right = dfs(preorder, inorder, preLeft + point - inLeft + 1, preRight, point + 1, inRight);
        }
        return root;
    }

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return dfs(preorder, inorder, 0, preorder.length - 1, 0, inorder.length - 1);
    }
}

File: core/ConstructBinaryTreeFromPreorderAndInorderTraversal/ConstructBinaryTreeFromPreorderAndInorderTraversal.py
from typing import List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def buildTree(preorder: List[int], inorder: List[int]) -> TreeNode:
    mid, point, length = preorder[0], 0, len(inorder)
    root = TreeNode(mid)
    while inorder[point] != mid:
        point += 1
    if point > 0:
        root.left = buildTree(preorder[1:point + 1], inorder[:point])
    if point < length - 1:
        root.right = buildTree(preorder[point + 1:], inorder[point + 1:])

    return root


